Enforce (on-chain) that Hydra transitions are authenticated by participants

KtorZ opened the pull request:
Fixes #284
---

@abailly-iohk There's currently one failing test in `hydra-cluster`

```
--match "/Test.LocalCluster/should produce blocks, provide funds, and send Hydra OCV transactions/"
```

but, I do think that the failing assertion on this test does really not have its place here. I left a comment in the last commit, I let you judge of that :blush: 

---

- :round_pushpin: **Inject signer mutation for CollectCom #284**
  
- :round_pushpin: **Sketch validation of CollectCom signers #284**
  
- :round_pushpin: **Extract tokens from outputs and implement signatory verification.**
  
- :round_pushpin: **Try to generate key for signing CollectCom tx #284**
  
- :round_pushpin: **Extract 'signWith' from TinyWallet and use it to authenticate Collectcom with party's credentials.**
  
- :round_pushpin: **Add signer's vk to the explicit required signers list when signing.**
  
- :round_pushpin: **Add 'mutate required signer' to Abort. Mutation tests fail.**
  
- :round_pushpin: **Refactor head contract to re-use logic for checking signatories on-chain.**
  
- :round_pushpin: **Add signers to the health abort transaction**
    Moved some logic around to make it possible.

- :round_pushpin: **Add mutation mutation to close tx about required signers.**
  
- :round_pushpin: **Implement signature verification on-chain for close transition.**
  
- :round_pushpin: **Repair healthy close tx, and correctly gather signatures on-chain.**
  
- :round_pushpin: **Fix finding Head currency symbol.**
    The initial version of the function was fine for pre-CollectCom transitions, but after the collect-com has happened, PTs are all aggregated under one policy and thus, there may be many assets locked by the head script.

- :round_pushpin: **Provide required signers when constructing Hydra on-chain transactions.**
    The rationale for doing this is double:

  (a) So far, we've been deferring signing up to inside the wallet and,
  requiring signature to happen on the transaction templates from
  Direct.Tx breaks a bit this abstraction. Moreso, we initially went
  that way because we thought that signing a transaction would result in
  the signatories being available on-chain in the Plutus context. Turns
  out that only explicitly specified signatories (in requiredSigners)
  end up in the Plutus context after all.

  (b) The 'State' abstraction we use internally is already tailored to
  one party and hold its 'ownVerificationKey'. So the Cardano signer key
  is already readily available and it is therefore much easier to
  provide it along the way. As a nice benefits, it makes it also clear
  from the signature of the low-level Direct.Tx functions that they now
  require to be explictly signed by a party.

- :round_pushpin: **Actually sign transaction AFTER modifying its body content**
    :grimacing: ... now that're changing the 'requiredSigners', we need to make sure to sign the modified body, otherwise, we produce an invalid witness.

- :round_pushpin: **Revert adding required signers during 'signWith'**
    This is in fact no longer needed and had become error-prone. Now that the Direct.Tx primitives take care of adding the required signer, there's no need to fiddle with the tx body during signing anymore.

- :round_pushpin: **Leave note about failing (albeit wrong) integration test.**

abailly committed:
Inject signer mutation for CollectCom #284

abailly committed:
Sketch validation of CollectCom signers #284

KtorZ committed:
Extract tokens from outputs and implement signatory verification.

abailly committed:
Try to generate key for signing CollectCom tx #284

KtorZ committed:
Extract 'signWith' from TinyWallet and use it to authenticate Collectcom with party's credentials.

KtorZ committed:
Add signer's vk to the explicit required signers list when signing.

KtorZ committed:
Add 'mutate required signer' to Abort. Mutation tests fail.

KtorZ committed:
Refactor head contract to re-use logic for checking signatories on-chain.

KtorZ committed:
Add signers to the health abort transaction

  Moved some logic around to make it possible.

KtorZ committed:
Add mutation mutation to close tx about required signers.

KtorZ committed:
Implement signature verification on-chain for close transition.

KtorZ committed:
Repair healthy close tx, and correctly gather signatures on-chain.

KtorZ committed:
Fix finding Head currency symbol.

  The initial version of the function was fine for pre-CollectCom transitions, but after the collect-com has happened, PTs are all aggregated under one policy and thus, there may be many assets locked by the head script.

KtorZ committed:
Provide required signers when constructing Hydra on-chain transactions.

  The rationale for doing this is double:

  (a) So far, we've been deferring signing up to inside the wallet and,
  requiring signature to happen on the transaction templates from
  Direct.Tx breaks a bit this abstraction. Moreso, we initially went
  that way because we thought that signing a transaction would result in
  the signatories being available on-chain in the Plutus context. Turns
  out that only explicitly specified signatories (in requiredSigners)
  end up in the Plutus context after all.

  (b) The 'State' abstraction we use internally is already tailored to
  one party and hold its 'ownVerificationKey'. So the Cardano signer key
  is already readily available and it is therefore much easier to
  provide it along the way. As a nice benefits, it makes it also clear
  from the signature of the low-level Direct.Tx functions that they now
  require to be explictly signed by a party.

KtorZ committed:
Actually sign transaction AFTER modifying its body content

  :grimacing: ... now that're changing the 'requiredSigners', we need to make sure to sign the modified body, otherwise, we produce an invalid witness.

KtorZ committed:
Revert adding required signers during 'signWith'

  This is in fact no longer needed and had become error-prone. Now that the Direct.Tx primitives take care of adding the required signer, there's no need to fiddle with the tx body during signing anymore.

KtorZ committed:
Leave note about failing (albeit wrong) integration test.

ghost commented:
@KtorZ There's even the following comment in the code:
```
-- TODO(AB): remove when DirectChainSpec is refactored to use cardano-api
failAfter 30 $ assertCanCallInitAndAbort cluster
```
This was something I did back in October when you were busy with wallet stuff we had barely started work on Direct tx submission, and I wanted to better understand how txs were built and submitted through the API.

abailly committed:
Simplify Cardano cluster integration test #284

* Rename file to match what we are testing, eg. CardanoCluster and not
LocalCluster
* remove failing assertion that's not relevant anymore: This used to
test we could construct basic Hydra transactions "by hand" but it's
getting more complicated and pretty much useless as it's thoroughly
tested elsewhere
